resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayer.java

```diff
@@ -398,107 +416,106 @@ private void saveSerialisedFormat(final String serialisedFormat, final LeafResou
      * Get a single resource by resource id with all parents resolved - not necessarily a leaf
      *
      * @param resourceId the id of the resource to get
-     * @return Optional.of a {@link Resource} in persistence, Optional.empty if not found
+     * @return {@link FunctionalIterator} of a {@link Resource} in persistence, empty {@link FunctionalIterator} if not found
      */
-    private Optional<Resource> getResourceById(final String resourceId) {
+    private FunctionalIterator<LeafResource> getResourceById(final String resourceId) {
         // Get resource entity from db
-        return resourceRepository.findByResourceId(resourceId)
-                // Get resource from db entity
-                .map(ResourceEntity::getResource)
-                // Resolve this resource's parents until no more parents found in db
-                .map(this::resolveParents);
+        FunctionalIterator<LeafResource> resourceIterator = FunctionalIterator.fromIterator(Collections.emptyIterator());
+        FunctionalIterator<ResourceEntity> entityIterator = FunctionalIterator.fromIterator(resourceRepository.findByResourceId(resourceId).iterator());
+        while (entityIterator.hasNext()) {
+            ResourceEntity entity = entityIterator.next();
+            resourceIterator = collectLeaves(resolveParents(entity.getResource()));
+        }
+        return resourceIterator;
     }
 
     // ~~~ Actual method implementations/overrides for PersistenceLayer interface ~~~ //
 
     // Given a resource, return all leaf resources underneath it with all parents resolved
     @Override
-    public Optional<Stream<LeafResource>> getResourcesById(final String resourceId) {
+    public FunctionalIterator<LeafResource> getResourcesById(final String resourceId) {
         LOGGER.debug("Getting resources by id '{}'", resourceId);
+        Iterator<LeafResource> resourceIterator = Collections.emptyIterator();
```

Comment:
You still need the optional, you can't just return an empty iterator.
The behaviour we want out of it is differentiating a "cache-hit" for an empty directory (`Optional.of(empty iterator)`) and a "cache-miss" (`Optional.empty()`)

If this has passed the tests, we need better tests

Reply:
The functionality has been updated to return an empty `Optional` if there is a cache-miss